File: android/ui-framework-sample/ui-framework/src/main/java/kh/mcnc/lib/components/BizKhmerCalendar.kt
package kh.mcnc.lib.components

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.ExperimentalAnimationApi
import androidx.compose.animation.SizeTransform
import androidx.compose.animation.core.tween
import androidx.compose.animation.slideInHorizontally
import androidx.compose.animation.slideOutHorizontally
import androidx.compose.animation.with
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ChevronLeft
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.Today
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kh.mcnc.lib.utils.CalendarConfig
import kh.mcnc.lib.utils.CalendarDay
import kh.mcnc.lib.utils.CalendarState
import kh.mcnc.lib.utils.SelectionMode
import kh.mcnc.lib.utils.IKhmerDate
import kh.mcnc.lib.utils.KhmerDate
import java.util.Calendar
import java.util.Date

@OptIn(ExperimentalAnimationApi::class)
@Composable
fun BizKhmerCalendar(
    config: CalendarConfig = CalendarConfig(),
    initialDate: Date = Date(),
    showActionButtons: Boolean = false,
    onDateSelected: ((date: Date, khmerDate: IKhmerDate?, fullText: String?) -> Unit)? = null,
    onRangeSelected: ((start: Date?, startKhmerDate: IKhmerDate?, startFullText: String?, end: Date?, endKhmerDate: IKhmerDate?, endFullText: String?) -> Unit)? = null,
    onMonthChanged: ((year: Int, month: Int) -> Unit)? = null,
    onCancel: (() -> Unit)? = null,
    onConfirm: ((start: Date?, end: Date?) -> Unit)? = null
) {
    val primaryColor = Color(0xFF1A2A5E)
    val dangerColor = Color(0xFFFF4D4F)
    val subtitleColor = primaryColor.copy(alpha = 0.7f)
    val rangeBgColor = dangerColor.copy(alpha = 0.1f)
    
    var currentMonth by remember(initialDate) { mutableStateOf(Calendar.getInstance().apply { time = initialDate }.get(Calendar.MONTH) + 1) }
    var currentYear by remember(initialDate) { mutableStateOf(Calendar.getInstance().apply { time = initialDate }.get(Calendar.YEAR)) }
    
    var selectedDate by remember(initialDate) { mutableStateOf<Date?>(initialDate) }
    var rangeStartDate by remember { mutableStateOf<Date?>(null) }
    var rangeEndDate by remember { mutableStateOf<Date?>(null) }
    
    val calendarState = remember(config) { CalendarState(config) }
    var days by remember { mutableStateOf(emptyList<CalendarDay>()) }
    var isLoading by remember { mutableStateOf(false) }
    var firstDayKhmer by remember { mutableStateOf<IKhmerDate?>(null) }
    
    var transitionDirection by remember { mutableStateOf(1) }
    
    val monthNames = listOf(
        "មករា", "កុម្ភៈ", "មីនា", "មេសា", "ឧសភា", "មិថុនា", 
        "កក្កដា", "សីហា", "កញ្ញា", "តុលា", "វិច្ឆិកា", "ធ្នូ"
    )
    
    val daysOfWeek = if (config.firstDayOfWeek == 1) {
        listOf("ចន្ទ", "អង្គារ", "ពុធ", "ព្រហ.", "សុក្រ", "សៅរ៍", "អាទិត្យ")
    } else {
        listOf("អាទិត្យ", "ចន្ទ", "អង្គារ", "ពុធ", "ព្រហ.", "សុក្រ", "សៅរ៍")
    }

    LaunchedEffect(currentMonth, currentYear, calendarState) {
        isLoading = true
        kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.Default) {
            val newDays = calendarState.generateDaysForMonth(currentYear, currentMonth)
            val newFirstDayKhmer = KhmerDate(1, currentMonth, currentYear).toLunar()
            days = newDays
            firstDayKhmer = newFirstDayKhmer
        }
        isLoading = false
        onMonthChanged?.invoke(currentYear, currentMonth)
    }

    fun isSameDay(d1: Date?, d2: Date?): Boolean {
        if (d1 == null || d2 == null) return false
        val cal1 = Calendar.getInstance().apply { time = d1 }
        val cal2 = Calendar.getInstance().apply { time = d2 }
        return cal1.get(Calendar.YEAR) == cal2.get(Calendar.YEAR) &&
               cal1.get(Calendar.MONTH) == cal2.get(Calendar.MONTH) &&
               cal1.get(Calendar.DAY_OF_MONTH) == cal2.get(Calendar.DAY_OF_MONTH)
    }

    fun getFullKhmerString(k: IKhmerDate): String {
        return "${k.lunarDay} ${k.lunarMonth} ឆ្នាំ${k.zodiacYear} ${k.stem} ព.ស. ${k.lunarYear}"
    }

    fun nextMonth() {
        transitionDirection = 1
        if (currentMonth == 12) {
            currentMonth = 1
            currentYear += 1
        } else {
            currentMonth += 1
        }
    }

    fun prevMonth() {
        transitionDirection = -1
        if (currentMonth == 1) {
            currentMonth = 12
            currentYear -= 1
        } else {
            currentMonth -= 1
        }
    }

    fun goToToday() {
        transitionDirection = 1
        val today = Calendar.getInstance()
        currentMonth = today.get(Calendar.MONTH) + 1
        currentYear = today.get(Calendar.YEAR)
    }

    fun onDayClick(day: CalendarDay) {
        if (!day.isEnabled || !day.isCurrentMonth) return
        
        when (config.selectionMode) {
            SelectionMode.Single -> {
                selectedDate = day.date
                onDateSelected?.invoke(day.date, day.khmerDate, getFullKhmerString(day.khmerDate))
            }
            SelectionMode.Week -> {
                val index = days.indexOfFirst { isSameDay(it.date, day.date) }
                if (index != -1) {
                    val rowStart = (index / 7) * 7
                    val weekDays = days.subList(rowStart, Math.min(rowStart + 7, days.size)).filter { it.isEnabled && !it.isHidden }
                    if (weekDays.isNotEmpty()) {
                        rangeStartDate = weekDays.first().date
                        rangeEndDate = weekDays.last().date
                        onRangeSelected?.invoke(
                            rangeStartDate, weekDays.first().khmerDate, getFullKhmerString(weekDays.first().khmerDate),
                            rangeEndDate, weekDays.last().khmerDate, getFullKhmerString(weekDays.last().khmerDate)
                        )
                    }
                }
            }
            SelectionMode.Range -> {
                if (rangeStartDate == null || (rangeStartDate != null && rangeEndDate != null)) {
                    rangeStartDate = day.date
                    rangeEndDate = null
                } else if (rangeStartDate != null && rangeEndDate == null) {
                    if (day.date.time < rangeStartDate!!.time) {
                        rangeStartDate = day.date
                    } else {
                        rangeEndDate = day.date
                        val startDay = days.find { isSameDay(it.date, rangeStartDate) }
                        onRangeSelected?.invoke(
                            rangeStartDate, startDay?.khmerDate, startDay?.let { getFullKhmerString(it.khmerDate) },
                            rangeEndDate, day.khmerDate, getFullKhmerString(day.khmerDate)
                        )
                    }
                }
            }
        }
    }

    val selectedKhmerDateFull = remember(selectedDate, config.selectionMode, days) {
        if (config.selectionMode == SelectionMode.Single && selectedDate != null) {
            val d = days.find { isSameDay(it.date, selectedDate) }
            if (d != null) getFullKhmerString(d.khmerDate) else null
        } else null
    }

    Column(modifier = Modifier
        .fillMaxWidth()
        .background(Color.White)
        .padding(16.dp)) {
        
        // Header
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 16.dp, bottom = 8.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = "${monthNames[currentMonth - 1]} $currentYear",
                fontSize = 18.sp,
                fontWeight = FontWeight.SemiBold,
                color = primaryColor,
                modifier = Modifier.clickable { /* TODO: Show Month Picker */ }
            )
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                IconButton(onClick = { prevMonth() }) {
                    Icon(Icons.Default.ChevronLeft, contentDescription = "Previous Month", tint = primaryColor)
                }
                IconButton(onClick = { goToToday() }) {
                    Icon(Icons.Default.Today, contentDescription = "Today", tint = primaryColor)
                }
                IconButton(onClick = { nextMonth() }) {
                    Icon(Icons.Default.ChevronRight, contentDescription = "Next Month", tint = primaryColor)
                }
            }
        }

        // Subtitle
        if (selectedKhmerDateFull != null) {
            Text(selectedKhmerDateFull, fontSize = 14.sp, color = subtitleColor, modifier = Modifier.padding(bottom = 16.dp))
        } else if (firstDayKhmer != null) {
            Text("ឆ្នាំ${firstDayKhmer!!.zodiacYear} ${firstDayKhmer!!.stem} ព.ស. ${firstDayKhmer!!.lunarYear}", fontSize = 14.sp, color = subtitleColor, modifier = Modifier.padding(bottom = 16.dp))
        } else {
            Spacer(modifier = Modifier.height(20.dp).padding(bottom = 16.dp))
        }

        // Weekdays
        Row(modifier = Modifier.fillMaxWidth().padding(bottom = 8.dp)) {
            daysOfWeek.forEachIndexed { index, dayName ->
                val isSunday = if (config.firstDayOfWeek == 1) index == 6 else index == 0
                Text(
                    text = dayName,
                    modifier = Modifier.weight(1f),
                    textAlign = TextAlign.Center,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = if (isSunday) dangerColor else primaryColor
                )
            }
        }

        // Calendar Grid
        AnimatedContent(
            targetState = "$currentYear-$currentMonth",
            transitionSpec = {
                if (transitionDirection > 0) {
                    slideInHorizontally(animationSpec = tween(300)) { width -> width } with
                    slideOutHorizontally(animationSpec = tween(300)) { width -> -width }
                } else {
                    slideInHorizontally(animationSpec = tween(300)) { width -> -width } with
                    slideOutHorizontally(animationSpec = tween(300)) { width -> width }
                }.using(SizeTransform(clip = false))
            }
        ) { _ ->
            Column(modifier = Modifier.fillMaxWidth()) {
                if (isLoading) {
                    val chunks = (0 until 42).chunked(7)
                    chunks.forEach { week ->
                        Row(modifier = Modifier.fillMaxWidth()) {
                            week.forEach { _ ->
                                Box(
                                    modifier = Modifier
                                        .weight(1f)
                                        .height(64.dp)
                                        .padding(vertical = 2.dp)
                                ) {
                                    BizSkeletonText(
                                        modifier = Modifier
                                            .fillMaxSize()
                                            .padding(horizontal = 2.dp)
                                            .clip(RoundedCornerShape(4.dp))
                                    )
                                }
                            }
                        }
                    }
                } else {
                    val chunks = days.chunked(7)
                    chunks.forEach { week ->
                        Row(modifier = Modifier.fillMaxWidth()) {
                            week.forEach { day ->
                                val isSelected = when (config.selectionMode) {
                                    SelectionMode.Single -> isSameDay(selectedDate, day.date)
                                    else -> isSameDay(rangeStartDate, day.date) || isSameDay(rangeEndDate, day.date)
                                }
                                val isRangeStart = config.selectionMode != SelectionMode.Single && isSameDay(rangeStartDate, day.date)
                                val isRangeEnd = config.selectionMode != SelectionMode.Single && isSameDay(rangeEndDate, day.date)
                                val isInRange = config.selectionMode != SelectionMode.Single && rangeStartDate != null && rangeEndDate != null &&
                                        day.date.time > rangeStartDate!!.time && day.date.time < rangeEndDate!!.time

                                Box(
                                    modifier = Modifier
                                        .weight(1f)
                                        .height(64.dp)
                                        .padding(vertical = 2.dp)
                                        .clickable(enabled = day.isEnabled && day.isCurrentMonth) { onDayClick(day) },
                                    contentAlignment = Alignment.Center
                                ) {
                                    if (!day.isHidden) {
                                        // Range backgrounds
                                        if (isInRange) {
                                            Box(modifier = Modifier.fillMaxSize().background(rangeBgColor))
                                        }
                                        if (isRangeStart && rangeEndDate != null && !isSameDay(rangeStartDate, rangeEndDate)) {
                                            Box(modifier = Modifier.fillMaxSize().padding(start = 32.dp).background(rangeBgColor))
                                        }
                                        if (isRangeEnd && rangeStartDate != null && !isSameDay(rangeStartDate, rangeEndDate)) {
                                            Box(modifier = Modifier.fillMaxSize().padding(end = 32.dp).background(rangeBgColor))
                                        }

                                        // Day content background
                                        val contentModifier = Modifier
                                            .fillMaxSize()
                                            .padding(horizontal = 2.dp)
                                            .clip(RoundedCornerShape(4.dp))
                                            .then(
                                                if (isSelected) Modifier.background(dangerColor)
                                                else if (day.isToday && !isSelected) Modifier.border(1.dp, dangerColor, RoundedCornerShape(4.dp))
                                                else Modifier
                                            )
                                            
                                        Box(modifier = contentModifier, contentAlignment = Alignment.Center) {
                                            if (day.isSpecial) {
                                                Box(modifier = Modifier.align(Alignment.TopEnd).padding(4.dp).size(6.dp).background(Color(0xFFFFA000), CircleShape))
                                            }
                                            
                                            Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.Center) {
                                                val textColor = if (isSelected) Color.White 
                                                        else if (!day.isEnabled || !day.isCurrentMonth) primaryColor.copy(alpha = 0.3f)
                                                        else if (day.isSunday) dangerColor 
                                                        else primaryColor
                                                val subTextColor = if (isSelected) Color.White else subtitleColor
                                                
                                                Text(
                                                    text = day.khmerDate.lunarDay,
                                                    fontSize = 8.sp,
                                                    color = subTextColor,
                                                    maxLines = 1,
                                                    modifier = Modifier.padding(top = 4.dp)
                                                )
                                                
                                                Text(
                                                    text = day.gregorianDay.toString(),
                                                    fontSize = 18.sp,
                                                    color = textColor
                                                )
                                                
                                                Text(
                                                    text = day.khmerDate.lunarMonth,
                                                    fontSize = 8.sp,
                                                    color = subTextColor,
                                                    maxLines = 1,
                                                    modifier = Modifier.padding(bottom = 4.dp)
                                                )
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        if (showActionButtons) {
            Row(
                modifier = Modifier.fillMaxWidth().padding(top = 16.dp),
                horizontalArrangement = Arrangement.End
            ) {
                TextButton(onClick = { onCancel?.invoke() }) {
                    Text("Cancel", color = primaryColor, fontWeight = FontWeight.SemiBold)
                }
                Spacer(modifier = Modifier.width(16.dp))
                Button(
                    onClick = { onConfirm?.invoke(if (config.selectionMode == SelectionMode.Single) selectedDate else rangeStartDate, if (config.selectionMode == SelectionMode.Single) null else rangeEndDate) },
                    colors = ButtonDefaults.buttonColors(containerColor = dangerColor)
                ) {
                    Text("Confirm", color = Color.White, fontWeight = FontWeight.SemiBold)
                }
            }
        }
    }
}

File: ios/ui-framework-sample/UIFramework/Sources/UIFramework/BizKhmerCalendar.swift
import SwiftUI

public struct KhmerCalendarSelection {
    public let date: Date
    public let khmerDate: IKhmerDate
    public let fullText: String
}

public struct BizKhmerCalendar: View {
    public var config: CalendarConfig
    @State public var initialDate: Date
    public var showActionButtons: Bool
    
    public var onDateSelected: ((KhmerCalendarSelection) -> Void)?
    public var onRangeSelected: ((KhmerCalendarSelection?, KhmerCalendarSelection?) -> Void)?
    public var onMonthChanged: ((Int, Int) -> Void)?
    public var onConfirm: ((KhmerCalendarSelection?, KhmerCalendarSelection?) -> Void)?
    public var onCancel: (() -> Void)?
    
    @State private var selectedDate: Date?
    @State private var rangeStartDate: Date?
    @State private var rangeEndDate: Date?
    
    @State private var currentMonth: Int
    @State private var currentYear: Int
    @State private var showMonthPicker = false
    
    @State private var dragOffset: CGFloat = 0
    @State private var isLoading: Bool = false
    
    @State private var days: [CalendarDay] = []
    @State private var firstDayKhmer: IKhmerDate?
    
    private let calendarState: CalendarState
    private let monthNames = ["មករា", "កុម្ភៈ", "មីនា", "មេសា", "ឧសភា", "មិថុនា", "កក្កដា", "សីហា", "កញ្ញា", "តុលា", "វិច្ឆិកា", "ធ្នូ"]
    
    public init(
        config: CalendarConfig = CalendarConfig(),
        initialDate: Date = Date(),
        showActionButtons: Bool = false,
        onDateSelected: ((KhmerCalendarSelection) -> Void)? = nil,
        onRangeSelected: ((KhmerCalendarSelection?, KhmerCalendarSelection?) -> Void)? = nil,
        onMonthChanged: ((Int, Int) -> Void)? = nil,
        onConfirm: ((KhmerCalendarSelection?, KhmerCalendarSelection?) -> Void)? = nil,
        onCancel: (() -> Void)? = nil
    ) {
        self.config = config
        self._initialDate = State(initialValue: initialDate)
        self.showActionButtons = showActionButtons
        
        self.onDateSelected = onDateSelected
        self.onRangeSelected = onRangeSelected
        self.onMonthChanged = onMonthChanged
        self.onConfirm = onConfirm
        self.onCancel = onCancel
        
        let cal = Calendar.current
        self._currentMonth = State(initialValue: cal.component(.month, from: initialDate))
        self._currentYear = State(initialValue: cal.component(.year, from: initialDate))
        
        self.calendarState = CalendarState(config: config)
        
        self._selectedDate = State(initialValue: initialDate)
    }
    
    private func getFullKhmerString(k: IKhmerDate) -> String {
        return "\(k.lunarDay) \(k.lunarMonth) ឆ្នាំ\(k.zodiacYear) \(k.stem) ព.ស. \(k.lunarYear)"
    }
    
    private var selectedKhmerDateFull: String? {
        if config.selectionMode == .single, let sel = selectedDate, let day = days.first(where: { isSameDay($0.date, sel) }) {
            return getFullKhmerString(k: day.khmerDate)
        }
        return nil
    }
    
    private var daysOfWeek: [String] {
        let firstDay = config.firstDayOfWeek
        if firstDay == 1 {
            return ["ចន្ទ", "អង្គារ", "ពុធ", "ព្រហ.", "សុក្រ", "សៅរ៍", "អាទិត្យ"]
        }
        return ["អាទិត្យ", "ចន្ទ", "អង្គារ", "ពុធ", "ព្រហ.", "សុក្រ", "សៅរ៍"]
    }
    
    private func isSundayIndex(_ idx: Int) -> Bool {
        return config.firstDayOfWeek == 1 ? idx == 6 : idx == 0
    }
    
    public var body: some View {
        VStack(spacing: 0) {
            headerView
            
            subtitleView
            
            weekdaysView
            
            calendarGridView
            
            if showActionButtons {
                actionsView
            }
        }
        .padding(16)
        .background(Color.white)
        .onAppear {
            updateCalendar()
        }
        .onChange(of: currentMonth) { _ in updateCalendar() }
        .onChange(of: currentYear) { _ in updateCalendar() }
        .onChange(of: config) { _ in updateCalendar() }
        .sheet(isPresented: $showMonthPicker) {
            MonthYearPicker(
                month: $currentMonth,
                year: $currentYear,
                isPresented: $showMonthPicker,
                monthNames: monthNames
            )
            .withPresentationDetents()
        }
    }
    
    private var headerView: some View {
        HStack {
            Button(action: { showMonthPicker = true }) {
                HStack(spacing: 4) {
                    Text("\(monthNames[currentMonth - 1]) \(String(currentYear))")
                        .font(.system(size: 18, weight: .semibold))
                        .foregroundColor(Color(red: 26/255, green: 42/255, blue: 94/255))
                    Image(systemName: "chevron.down")
                        .font(.system(size: 14))
                }
            }
            Spacer()
            HStack(spacing: 8) {
                Button(action: { animateToPrevMonth() }) {
                    Image(systemName: "chevron.left")
                        .frame(width: 24, height: 24)
                }
                Button(action: goToToday) {
                    Image(systemName: "calendar")
                        .frame(width: 24, height: 24)
                        .foregroundColor(Color(red: 0, green: 51/255, blue: 153/255))
                }
                Button(action: { animateToNextMonth() }) {
                    Image(systemName: "chevron.right")
                        .frame(width: 24, height: 24)
                }
            }
            .foregroundColor(.primary)
        }
        .padding(.vertical, 8)
    }
    
    @ViewBuilder
    private var subtitleView: some View {
        if let subtitle = selectedKhmerDateFull {
            Text(subtitle)
                .font(.system(size: 14))
                .foregroundColor(Color(red: 26/255, green: 42/255, blue: 94/255, opacity: 0.7))
                .padding(.bottom, 16)
        } else if let fk = firstDayKhmer {
            Text("ឆ្នាំ\(fk.zodiacYear) \(fk.stem) ព.ស. \(fk.lunarYear)")
                .font(.system(size: 14))
                .foregroundColor(Color(red: 26/255, green: 42/255, blue: 94/255, opacity: 0.7))
                .padding(.bottom, 16)
        } else {
            Spacer().frame(height: 36) // 14 font size approx height with padding
        }
    }
    
    private var weekdaysView: some View {
        HStack(spacing: 0) {
            ForEach(0..<7, id: \.self) { idx in
                Text(daysOfWeek[idx])
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundColor(isSundayIndex(idx) ? Color.red : Color(red: 26/255, green: 42/255, blue: 94/255))
                    .frame(maxWidth: .infinity)
            }
        }
        .padding(.bottom, 8)
    }
    
    private var calendarGridView: some View {
        let columns = Array(repeating: GridItem(.flexible(), spacing: 0), count: 7)
        return LazyVGrid(columns: columns, spacing: 0) {
            if isLoading {
                ForEach(0..<42, id: \.self) { _ in
                    skeletonCell
                }
            } else {
                ForEach(days) { day in
                    dayCell(for: day)
                }
            }
        }
        .offset(x: dragOffset)
        .gesture(
            DragGesture()
                .onChanged { value in
                    dragOffset = value.translation.width
                }
                .onEnded { value in
                    if value.translation.width > 50 {
                        animateToPrevMonth()
                    } else if value.translation.width < -50 {
                        animateToNextMonth()
                    } else {
                        withAnimation { dragOffset = 0 }
                    }
                }
        )
    }
    
    private var skeletonCell: some View {
        ZStack {
            BizSkeletonText()
                .padding(2)
        }
        .frame(height: 64)
    }
    
    private func dayCell(for day: CalendarDay) -> some View {
        ZStack {
            if day.isHidden {
                Color.clear
            } else {
                let isDisabled = !day.isEnabled || !day.isCurrentMonth
                let selected = isSelected(day)
                let inRange = isInRange(day)
                let rangeStart = isRangeStart(day)
                let rangeEnd = isRangeEnd(day)
                
                ZStack {
                    if inRange || rangeStart || rangeEnd {
                        HStack(spacing: 0) {
                            if rangeStart && !rangeEnd {
                                Color.clear.frame(maxWidth: .infinity)
                                Color.red.opacity(0.1).frame(maxWidth: .infinity)
                            } else if rangeEnd && !rangeStart {
                                Color.red.opacity(0.1).frame(maxWidth: .infinity)
                                Color.clear.frame(maxWidth: .infinity)
                            } else if inRange {
                                Color.red.opacity(0.1)
                            } else {
                                Color.clear
                            }
                        }
                    }
                    
                    VStack(spacing: 0) {
                        Text(day.khmerDate.lunarDay)
                            .font(.system(size: 8))
                            .foregroundColor(selected ? .white : Color(red: 26/255, green: 42/255, blue: 94/255, opacity: 0.7))
                            .lineLimit(1)
                            .padding(.top, 4)
                        
                        Text("\(day.gregorianDay)")
                            .font(.system(size: 18))
                            .foregroundColor(
                                selected ? .white :
                                    (day.isSunday ? .red :
                                        (isDisabled ? Color.gray.opacity(0.5) : Color.primary))
                            )
                        
                        Text(day.khmerDate.lunarMonth)
                            .font(.system(size: 8))
                            .foregroundColor(selected ? .white : Color(red: 26/255, green: 42/255, blue: 94/255, opacity: 0.7))
                            .lineLimit(1)
                            .padding(.bottom, 4)
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                    .background(
                        selected ? Color.red : Color.clear
                    )
                    .cornerRadius(4)
                    .overlay(
                        RoundedRectangle(cornerRadius: 4)
                            .stroke(day.isToday && !selected ? Color.red : Color.clear, lineWidth: 1)
                    )
                    .overlay(
                        Group {
                            if day.isSpecial {
                                Circle()
                                    .fill(Color.orange)
                                    .frame(width: 4, height: 4)
                                    .position(x: 34, y: 6) // Approx position
                            }
                        }
                    )
                    .padding(2)
                }
                .frame(height: 64)
                .opacity(isDisabled ? 0.3 : 1.0)
                .onTapGesture {
                    onDayClick(day)
                }
            }
        }
        .frame(height: 64)
    }
    
    private var actionsView: some View {
        HStack(spacing: 16) {
            Spacer()
            Button("Cancel") {
                onCancel?()
            }
            .font(.system(size: 14, weight: .semibold))
            .foregroundColor(Color(red: 26/255, green: 42/255, blue: 94/255))
            
            Button("Confirm") {
                onConfirmAction()
            }
            .font(.system(size: 14, weight: .semibold))
            .foregroundColor(.white)
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
            .background(Color.red)
            .cornerRadius(4)
        }
        .padding(.top, 16)
    }
    
    private func updateCalendar() {
        isLoading = true
        let year = currentYear
        let month = currentMonth
        
        DispatchQueue.global(qos: .userInitiated).async {
            let newDays = self.calendarState.generateDaysForMonth(year: year, month: month)
            let fk = KhmerDate(day: 1, month: month, year: year).toLunar()
            
            DispatchQueue.main.async {
                self.days = newDays
                self.firstDayKhmer = fk
                self.isLoading = false
                self.onMonthChanged?(year, month)
            }
        }
    }
    
    private func animateToPrevMonth() {
        withAnimation(.easeInOut(duration: 0.2)) {
            dragOffset = UIScreen.main.bounds.width
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
            prevMonth()
            dragOffset = -UIScreen.main.bounds.width
            withAnimation(.easeInOut(duration: 0.2)) {
                dragOffset = 0
            }
        }
    }
    
    private func animateToNextMonth() {
        withAnimation(.easeInOut(duration: 0.2)) {
            dragOffset = -UIScreen.main.bounds.width
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
            nextMonth()
            dragOffset = UIScreen.main.bounds.width
            withAnimation(.easeInOut(duration: 0.2)) {
                dragOffset = 0
            }
        }
    }
    
    private func prevMonth() {
        if currentMonth == 1 {
            currentMonth = 12
            currentYear -= 1
        } else {
            currentMonth -= 1
        }
    }
    
    private func nextMonth() {
        if currentMonth == 12 {
            currentMonth = 1
            currentYear += 1
        } else {
            currentMonth += 1
        }
    }
    
    private func goToToday() {
        let cal = Calendar.current
        let today = Date()
        currentMonth = cal.component(.month, from: today)
        currentYear = cal.component(.year, from: today)
    }
    
    private func isSameDay(_ d1: Date?, _ d2: Date?) -> Bool {
        guard let d1 = d1, let d2 = d2 else { return false }
        let cal = Calendar.current
        return cal.isDate(d1, inSameDayAs: d2)
    }
    
    private func isSelected(_ day: CalendarDay) -> Bool {
        if config.selectionMode == .single {
            return isSameDay(selectedDate, day.date)
        }
        return isSameDay(rangeStartDate, day.date) || isSameDay(rangeEndDate, day.date)
    }
    
    private func isInRange(_ day: CalendarDay) -> Bool {
        guard config.selectionMode != .single, let start = rangeStartDate, let end = rangeEndDate else { return false }
        return day.date > start && day.date < end
    }
    
    private func isRangeStart(_ day: CalendarDay) -> Bool {
        if config.selectionMode == .single { return false }
        return isSameDay(rangeStartDate, day.date)
    }
    
    private func isRangeEnd(_ day: CalendarDay) -> Bool {
        if config.selectionMode == .single { return false }
        return isSameDay(rangeEndDate, day.date)
    }
    
    private func onDayClick(_ day: CalendarDay) {
        if !day.isEnabled || !day.isCurrentMonth { return }
        
        if config.selectionMode == .single {
            selectedDate = day.date
            onDateSelected?(KhmerCalendarSelection(date: day.date, khmerDate: day.khmerDate, fullText: getFullKhmerString(k: day.khmerDate)))
        } else if config.selectionMode == .week {
            if let index = days.firstIndex(where: { isSameDay($0.date, day.date) }) {
                let rowStart = (index / 7) * 7
                let weekDays = days[rowStart..<min(rowStart + 7, days.count)].filter { $0.isEnabled && !$0.isHidden }
                if let first = weekDays.first, let last = weekDays.last {
                    rangeStartDate = first.date
                    rangeEndDate = last.date
                    onRangeSelected?(
                        KhmerCalendarSelection(date: first.date, khmerDate: first.khmerDate, fullText: getFullKhmerString(k: first.khmerDate)),
                        KhmerCalendarSelection(date: last.date, khmerDate: last.khmerDate, fullText: getFullKhmerString(k: last.khmerDate))
                    )
                }
            }
        } else {
            if rangeStartDate == nil || (rangeStartDate != nil && rangeEndDate != nil) {
                rangeStartDate = day.date
                rangeEndDate = nil
            } else if let start = rangeStartDate, rangeEndDate == nil {
                if day.date < start {
                    rangeStartDate = day.date
                } else {
                    rangeEndDate = day.date
                    let startDay = days.first { isSameDay($0.date, start) }
                    onRangeSelected?(
                        startDay.map { KhmerCalendarSelection(date: $0.date, khmerDate: $0.khmerDate, fullText: getFullKhmerString(k: $0.khmerDate)) },
                        KhmerCalendarSelection(date: day.date, khmerDate: day.khmerDate, fullText: getFullKhmerString(k: day.khmerDate))
                    )
                }
            }
        }
    }
    
    private func onConfirmAction() {
        let startDay = days.first { isSameDay($0.date, config.selectionMode == .single ? selectedDate : rangeStartDate) }
        let endDay = days.first { isSameDay($0.date, rangeEndDate) }
        
        if config.selectionMode == .single {
            onConfirm?(
                startDay.map { KhmerCalendarSelection(date: $0.date, khmerDate: $0.khmerDate, fullText: getFullKhmerString(k: $0.khmerDate)) },
                nil
            )
        } else {
            onConfirm?(
                startDay.map { KhmerCalendarSelection(date: $0.date, khmerDate: $0.khmerDate, fullText: getFullKhmerString(k: $0.khmerDate)) },
                endDay.map { KhmerCalendarSelection(date: $0.date, khmerDate: $0.khmerDate, fullText: getFullKhmerString(k: $0.khmerDate)) }
            )
        }
    }
}
